[FLINK-11067][table] Convert TableEnvironments to interfaces

hequn8128 opened the pull request:
## What is the purpose of the change

This pull request converts TableEnvironments to interfaces.

## Brief change log

### TableEnvironment interface
There are 5 TableEnvironment interfaces:
```
// in table-api-java-bridge
org.apche.flink.table.api.java.BatchTableEnvironment
org.apche.flink.table.api.java.StreamTableEnvironment
// in table-api-scala-bridge
org.apche.flink.table.api.scala.BatchTableEnvironment
org.apche.flink.table.api.scala.StreamTableEnvironment
// in table-api-java 
org.apche.flink.table.api.TableEnvironment
```
Take `java.BatchTableEnvironment` as an example, it contains a static method `static StreamTableEnvironment create(StreamExecutionEnvironment executionEnvironment, TableConfig tableConfig)`. The `create` method uses reflection to create a TableEnvironment instance now. We can replace reflection strategy to discovery strategy when introducing Planner. 

### TableConfig
- Rename CalciteConfig to PlannerConfig.
- Add DefaultPlannerConfig in flink-table-planner

### BatchTableDescriptor & StreamTableDescriptor
- Port BatchTableDescriptor and StreamTableDescriptor to api-java, so that the connect() method in StreamTableEnvironment can return StreamTableDescriptor. This makes the api compitable.
- There are other classes that have to be ported into common, such as `FieldComputer`, `Rowtime`, `Schema`, etc. When porting `WatermarkStrategy`, it affects the result of serialization resulting in incompatibility. It seems there are no better ways to achive compatibility. 
- Convert `TableFactoryUtil` into interface as it is used in ConnectTableDescriptor which needs to be ported into api-java.

### other changes
- Remove duplicate comments in TableEnvImpl. We have moved these comments into the interface.
- This pr takes no consideration of blink. I opened another jira([FLINK-12016](https://issues.apache.org/jira/browse/FLINK-12016)) to address the problem of Blink. In the jira we need to rename the package name of blink, because a lot of classes was copied from flink directly which brings class conflict when we port TableEnvironment into api module.

## Verifying this change

This change is already covered by existing tests. Also `ScalaTableEnvironmentCreationValidationTest` and `JavaTableEnvironmentCreationValidationTest` have been added to test the creation of table environments.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

twalthr commented:
@flinkbot approve-until consensus

dawidwys commented:
Hi @hequn8128 Would you mind rebasing this commit on current master?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java:
```diff
@@ -0,0 +1,289 @@
+	 * @param position cursor position
+	 * @return completion hints that fit at the current cursor position
+	 */
+	String[] getCompletionHints(String statement, int position);
```
I wonder if we should have this method in TableEnvironment interface... It is useful only for the sql-client and only for SQL (it makes no sense for Table API). SQL is (or will be ;) ) handled exclusively by the planner (maybe + access to Catalog, which will be a separate entity from `TableEnvironment` anyway). TableEnvironment just passes the string to the planner. I think we can base the sql client on the Planner interface directly in the long run.  

Therefore I would be up for dropping this method from `TableEnvironment` interface. WDYT @hequn8128 @twalthr ?

hequn8128 replied:
Make sense. Maybe we can deprecate this method first and drop it in the next release, as it is public in the previous `TableEnvironment`. What do you think?

twalthr replied:
I think this method is useful for anyone that uses the Table API as a library to build a service around it. I don't have a strong opinion here, but I think this could also stay as part of the interface.

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java:
```diff
@@ -0,0 +1,289 @@
+	/**
+	 * Returns the {@link QueryConfig} depends on the concrete type of this TableEnvironment.
+	 */
+	QueryConfig queryConfig();
```
Why do we need this method in the `TableEnvironment` interface? This is not a property of a whole environment, but it is query specific. Therefore I think it does not belong here.

I think it is used as a default value for methods that do not expect `QueryConfig`, but it can be a private internal method, can't it?

hequn8128 replied:
Good catch. I think we can drop this method in the `TableEnvironment`.

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java:
```diff
@@ -0,0 +1,289 @@
+ *     <li>Registering a Table in the internal catalog</li>
+ *     <li>Registering an external catalog</li>
+ *     <li>Executing SQL queries</li>
+ *     <li>Registering a user-defined (scalar, table, or aggregation) function</li>
```
Could we update this comment to current state? In the base interface we cannot register table nor aggregation function. I know that this is a temporary thing, but maybe at least link to the implementations that can do that? Sth like: `Registering a user-defined scalar function and table or aggregation function (via concrete implementation)`

hequn8128 replied:
You are right, we can't register table or aggregate function with the `TableEnvironment` interface. I will correct the comments. Thank you.

bowenli86 replied:
Just curious, why we can't register table or aggregate function with the TableEnvironment interface now? Sounds like it's only a temporary thing?

bowenli86 replied:
Just curious, why we can't register table or aggregate function with the TableEnvironment interface now? Sounds like it's only a temporary thing?

dawidwys replied:
We still can register functions, just not on the unified Environment. You can still register them on `Stream/BatchTableEnvironment`, which are the only entry points. We will allow registering aggregate & table functions on `TableEnvironment` once we rework type system.

hequn8128 commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java:
```diff
@@ -0,0 +1,289 @@
+	 * @param fieldTypes The field types to register with the {@link TableSink}.
+	 * @param tableSink The {@link TableSink} to register.
+	 */
+	void registerTableSink(String name, String[] fieldNames, TypeInformation<?>[] fieldTypes, TableSink<?> tableSink);
```
I think you are right. However, this method may be widely used by users. How about deprecating it in this pr and remove it later?

dawidwys replied:
That's unfortunate as we are extracting new interface, but I think you are right. Let's deprecate it first.

hequn8128 commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableEnvironment.java:
```diff
@@ -0,0 +1,289 @@
+	 *
+	 * @return A list of the names of all registered tables.
+	 */
+	String[] listTables();
```
I think it only returns directly registered tables. Listing tables in Catalog can use CatalogAPIs. Does it make sense to you? Maybe we should make the comment more accurate.

bowenli86 replied:
In design of new catalog framework, tables - both Flink tables and external tables - will always belong to a database under a catalog. Moreover, Flink will always have a current default catalog and database (which users can change), for user's convenience. When we reach that phase, listTables() would return all registered tables within the default database under the default catalog.

We can update the comment more accurately once the above design is fully realized and put into production use (spoil alert - very soon :).

bowenli86 replied:
In design of new catalog framework, tables - both Flink tables and external tables - will always belong to a database under a catalog. Moreover, Flink will always have a current default catalog and database (which users can change), for user's convenience. When we reach that phase, listTables() would return all registered tables within the default database under the default catalog.

Once the above design is fully realized and put into production use, I will update the comment more accurately (spoil alert - very soon :).

hequn8128 commented:
@dawidwys  Thanks a lot for your review. I think you have given a lot of good suggestions!

1. As for the useless APIs in `TableEnvironment` interface, maybe we can deprecate first and remove them later. This can make our interface compitable. 

2. To uncouple the `Descriptor` with `TableEnvironment`,  we can change registerTableSource to getTableSource in `RegistrableDescriptor`(also should rename `RegistrableDescriptor`). Once we get the table source or sink, then we can register the table source or sink with TableEnvironment. Althougn this makes users impossible to register table source(or sink) directly, I am ok with it.

3. For the `TableFactoryUtil`, I am not sure if we can port it into api or common module(I will make sure later). However, I do find we can also uncouple TableEnvironment here. In the current implementation, it seems table env is only used to decide whether it is a batch or stream source(or sink). We can also use descriptor to decide, i.e., a StreamTableDescriptor means a stream source(or sink). 

For 2 and 3, I think the biggest obstacle is we may need to fix the problem in a backwards compatible way. @twalthr Would be great to also have your opinions here.

Best, Hequn

hequn8128 commented:
> Hi @hequn8128 Would you mind rebasing this commit on current master?

Sure :-)
I will do it ASAP while addressing the comments that are quit sure.

hequn8128 commented:
Address comments and rebase to master.

dawidwys commented:
Hi @hequn8128,
I had another look and some discussion with @twalthr about this PR and we came up with few ideas how to proceed further with this PR:

1. The `Stream/BatchTableDescriptor`, `Stream/BatchTableSource`, `Stream/BatchTableSink`, `Stream/BatchTableSourceFactory`, `Stream/BatchTableSinkFactory` should end up in the bridging module. In the long run we need to rework them anyway. The new versions will end up in the api-java.
This might solve our problems with the reflection lookup for factories.
2. Could you please split the PR into smaller commits that move a smaller "hierarchies" of classes one at a time. The last commit would move just extract the TableEnvironment once all the needed classes are already moved? This would help us review this PR tremendously.

hequn8128 commented:
@bowenli86 Thanks for your review and providing valuable information for the catalog.

- Re TableEnvironment.
Yes, you are right. Currently, we must use a java or scala Environment to register table or aggregate functions due to TypeInformations are handled differently between java and scala. However, I think this would be solved by FLIP-37.

- Re listTables.
Nice spoil alert. Looking forward to it. :-)

hequn8128 commented:
@dawidwys Thanks for your suggestions. I have split the PR and rebased to the master. Would be great if you can take a look.

dawidwys commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/calcite/DefaultPlannerConfig.scala:
```diff
@@ -27,22 +27,15 @@ import org.apache.calcite.sql.parser.SqlParser
-@deprecated(
-  "This method will be removed temporarily while the API is uncoupled from Calcite.",
-  "1.8.0")
-class CalciteConfigBuilder {
```
How about we leave the old name? I think the reason for renaming it was that in the future this config could evolve into more complex planner specific config, right? 

I think though we should try to modularize the config a bit, I think we could take the approach as in the comment on `PlannerConfig`

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/PlannerConfig.java:
```diff
@@ -0,0 +1,29 @@
+ * The {@link PlannerConfig} holds parameters to configure the behavior of queries.
+ */
+@PublicEvolving
+public interface PlannerConfig {
```
I was thinking we could try to modularize the `PlannerConfig` a bit, so that we don't need to pass along all the configuration everywhere.

We could take the example of calcite's `Wrapper` class here. And add an `unwrap` method to the `PlannerConfig`:
```
@PublicEvolving
public interface PlannerConfig {
  	@SuppressWarnings("unchecked")
	default <T extends PlannerConfig> Optional<T> unwrap(Class<T> type) {
		if (type.isInstance(this)) {
			return Optional.of((T) this);
		} else {
			return Optional.empty();
		}
	}
}
```

We could later use this method to modularize the `PlannerConfig` having sth like `CompositePlannerConfig` which would be a hashmap from class to its instance.

wuchong replied:
What about return T directly instead of wrapping in a `Optional`.  We can return null or throw exception if the type not matched. It will allow us to write more fluent code when we are clear what the `PlannerConfig` type is, i.e. `tableConfig.getPlannerConfig().unwrap(DefaultPlannerConfig.class)`.

hequn8128 replied:
As for Optional[T] or T, I'm fine with both. @dawidwys What's your opinions here?

dawidwys replied:
I am strongly for returning `Optional<T>`. This makes the contract clear. This is a user facing interface and you can never say for sure, that it contains required class. `Optional` makes it simple and clear how to handle default value: 

```
tableConfig.getPlannerConfig().unwrap(DefaultPlannerConfig.class).orElse(DefaultPlannerConfig.DEFAULT)
```

Of course you can still pass the unwrapped value internally, and unwrap it early in the stack. This way we can also modularize config. Imagine at some point we will have also e.g. `CodeGenerationConfig`. Than we can have:

```

def query(PlannerConfig config) {
   configureCalcite(config.unwarp(CalciteConfig.class).orElse(CalciteConfig.DEFAULT))
   configureCodeGen(config.unwarp(CodeGenConfig.class).orElse(CodeGenConfig.DEFAULT))
}

def configureCalcite(CalciteConfig config)

def configureCodeGeneration(CodeGenConfig config)
```

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableConfig.java:
```diff
@@ -0,0 +1,143 @@
+
+	/**
+	 * Returns the default context for decimal division calculation.
+	 * [[_root_.java.math.MathContext#DECIMAL128]] by default.
```
Update to `{@link` .

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableConfig.java:
```diff
@@ -0,0 +1,143 @@
+
+	/**
+	 * Sets the default context for decimal division calculation.
+	 * [[_root_.java.math.MathContext#DECIMAL128]] by default.
```
Update to `{@link` .

dawidwys commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/rules/datastream/DataStreamWindowJoinRule.scala:
```diff
@@ -80,7 +80,7 @@ class DataStreamWindowJoinRule
         leftRowSchema.arity,
         join.getRowType,
         join.getCluster.getRexBuilder,
-        TableConfig.DEFAULT)
```
I think we still can have `TableConfig.DEFAULT` instance, can't we?

hequn8128 replied:
The `DEFAULT` in scala is actually a method. We need to create each time because the tableConfig could be modified in anyplace. 
To make it more convenient, maybe we can add a static `getDefault()` method in `TableConfig`. What do you think?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/AmbiguousTableFactoryException.java:
```diff
@@ -0,0 +1,85 @@
+
+	@Override
+	public String getMessage() {
+		return "More than one suitable table factory for '" + factoryClass.getName() + "' could\n" +
```
Could we use `String.format()` for that?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/NoMatchingTableFactoryException.java:
```diff
@@ -0,0 +1,81 @@
+
+	@Override
+	public String getMessage() {
+		return "Could not find a suitable table factory for '" + factoryClass.getName() + "' in\n" +
```
use `String.format()`

dawidwys commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Rowtime.java:
```diff
@@ -0,0 +1,145 @@
+	 *                  from the physical type.
+	 */
+	public Rowtime timestampsFromExtractor(TimestampExtractor extractor) {
+		internalProperties.putString(ROWTIME_TIMESTAMPS_TYPE, ROWTIME_TIMESTAMPS_TYPE_VALUE_CUSTOM);
```
I think this changes the property format that we use underneath. How about we add method `TimestampExtractor#toProperties()`? The default implementation could be like this:

```
	protected Map<String, String> toProperties() {
		Map<String, String> properties = new HashMap<>();
		properties.put(ROWTIME_TIMESTAMPS_TYPE -> ROWTIME_TIMESTAMPS_TYPE_VALUE_CUSTOM);
		properties.put(ROWTIME_TIMESTAMPS_TYPE -> extractor.getClass.getName);
		properties.put(ROWTIME_TIMESTAMPS_TYPE -> EncodingUtils.encodeObjectToString(extractor));
		return properties;
	}
```

In the classes that used custom format (e.g. ExisitngField) we can still use more specific properties.

hequn8128 replied:
Good idea!

hequn8128 replied:
Sorry, I forgot this. I will try my best not to happen this again.

dawidwys commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/descriptors/RowtimeValidator.scala:
```diff
@@ -100,23 +100,33 @@ class RowtimeValidator(
+  @Deprecated val ROWTIME_WATERMARKS_DELAY = Rowtime.ROWTIME_WATERMARKS_DELAY
 
   // utilities
 
```
There are unused methods in this class, e.g. `normalizeTimestampExtractor`.

wuchong commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/TableEnvImpl.scala:
```diff
@@ -65,16 +58,20 @@ import org.apache.flink.table.typeutils.TimeIndicatorTypeInfo
+
+  // Init PlannerConfig with the DefaultPlannerConfig
+  if (config.getPlannerConfig == null) {
+    config.setPlannerConfig(DefaultPlannerConfig.createBuilder().build())
```
```suggestion
    config.setPlannerConfig(DefaultPlannerConfig.DEFAULT)
```

use `DefaultPlannerConfig.DEFAULT` instead ?

dawidwys replied:
Let's not set anything here. `TableEnvImpl` is meant to be moved to api module. `DefaultPlannerConfig` won't be in the scope anymore.

This is where `Optional` comes in handy. On the planner side we will just do: 

```
plannerConfig.unwrap(DefaultPlannerConfig.class).orElse(DefaultPlannerConfig.DEFAULT)
```

hequn8128 commented:
@dawidwys Hi, I have addressed all your comments and updated the PR. Thanks a lot for your great suggestions. 

Best, Hequn

dawidwys commented:
3e4abdb2896ed136a7cb5ce2a4149afd72927816 lgtm

dawidwys commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/sources/tsextractors/ExistingField.scala:
```diff
@@ -75,4 +78,11 @@ final class ExistingField(val field: String) extends TimestampExtractor {
+  override def toProperties: util.Map[String, String] = {
+    val javaMap = new util.HashMap[String, String]()
+    javaMap.put(Rowtime.ROWTIME_TIMESTAMPS_TYPE, Rowtime.ROWTIME_TIMESTAMPS_TYPE_VALUE_FROM_FIELD)
+    javaMap.put(Rowtime.ROWTIME_TIMESTAMPS_FROM, this.getArgumentFields.apply(0))
```
`javaMap.put(Rowtime.ROWTIME_TIMESTAMPS_FROM, field)`

hequn8128 replied:
Good catch!

dawidwys commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/descriptors/RowtimeTest.scala:
```diff
@@ -87,7 +87,7 @@ class RowtimeTest extends DescriptorTestBase {
         "B0b3JzLlJvd3RpbWVUZXN0JEN1c3RvbUFzc2lnbmVyeDcuDvfbu0kCAAB4cgBHb3JnLmFwYWNoZS5mbGluay" +
         "50YWJsZS5zb3VyY2VzLndtc3RyYXRlZ2llcy5QdW5jdHVhdGVkV2F0ZXJtYXJrQXNzaWduZXKBUc57oaWu9A" +
         "IAAHhyAD1vcmcuYXBhY2hlLmZsaW5rLnRhYmxlLnNvdXJjZXMud21zdHJhdGVnaWVzLldhdGVybWFya1N0cm" +
-        "F0ZWd5mB_uSxDZ8-MCAAB4cA")
```
What is the reason for this mismatch?

hequn8128 replied:
Previously, the `WatermarkStrategy` extends a Scala `Serializable`. But now it extends the Java Serializable. Also, we make it extends `Descriptor`. The class has been changed.
Same reason for `TimestampExtractor`.

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryService.java:
```diff
@@ -0,0 +1,363 @@
+public class TableFactoryService {
+
+	private static final ServiceLoader<TableFactory> defaultLoader = ServiceLoader.load(TableFactory.class);
+	private static final transient Logger LOG = LoggerFactory.getLogger(TableFactoryService.class);
```
Why is it `transient`? By default static fields won't be serialized unless you implement some custom weird logic. Either remove transient or add a comment why we need it.

hequn8128 replied:
You are right. There is no point in declaring a static member field as transient.

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryService.java:
```diff
@@ -0,0 +1,363 @@
+			}
+		});
+
+		AtomicReference<Optional<String>> lastKey = new AtomicReference<>(Optional.empty());
```
Let's rewrite this as for-loop if we need additional return from the loop. I had hard time understanding why do we need `AtomicReference` here.

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryService.java:
```diff
@@ -0,0 +1,363 @@
+			TableFactory factory = classFactories.get(0);
+			Tuple2<List<String>, List<String>> tuple2 = normalizeSupportedProperties(factory);
+
+			String errorMessage = "The matching factory '" + factory.getClass().getName() + "' doesn't support '" + lastKey.get().get() + "'.\n" +
```
Use `String.format()`

dawidwys commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/factories/TableFactoryUtil.java:
```diff
@@ -0,0 +1,137 @@
+		try {
+			if (descriptor instanceof BatchTableDescriptor || isBatchExternalCatalogTable(descriptor)) {
+				Object object = TableFactoryService.find(
+					Class.forName("org.apache.flink.table.factories.BatchTableSinkFactory"),
```
How about we introduce common interface `TableSink/SourceFactory` for Stream and Batch Factories? This interface would return always `TableSink/Source` not its Stream/Batch counterparts.

This way:

1. We wouldn't need the reflective lookup. We would just have:
```
TableSinkFactory factory = TableFactoryService.find(TableSinkFactory.class, properties);
return factory.createTableSink(...);
```
2. This would already be prepared for unifying Batch and Stream Sinks/Sources in the future.

hequn8128 replied:
This would be good if we change it to TableSink/Source. However, I'm concern about the compatible problem. These methods are widely used by users. @twalthr What's your opinions here.

dawidwys replied:
I am not saying we should change the signature of e.g.`BatchTableSourceFactory` What I am suggesting is the following:

```
interface TableSourceFactory {
   TableSource createTableSource(...)
}

trait BatchTableSourceFactory extends TableSourceFactory {
   def createBatchTableSink(...): BatchTableSink[T]

   override def createTableSource(...) : TableSource = createBatchTableSource(...)
}

trait StreamTableSourceFactory extends TableSourceFactory {
   def createStreamTableSource(...): StreamTableSource[T]

   override def createTableSource(...) : TableSource = createStreamTableSource(...)
}
```

hequn8128 replied:
Good idea! This is a nice solution.

hequn8128 commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/TableConfig.java:
```diff
@@ -0,0 +1,148 @@
+	/**
+	 * Returns the current configuration of Planner for Table API and SQL queries.
+	 */
+	public Optional<PlannerConfig> getPlannerConfig() {
```
Good idea!

dawidwys commented on flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/api/java/BatchTableEnvironment.java:
```diff
@@ -0,0 +1,289 @@
+	 *
+	 * Example:
+	 *
+	 * {{{
```
Convert to javadoc formatting.

dawidwys commented on flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/api/java/BatchTableEnvironment.java:
```diff
@@ -0,0 +1,289 @@
+	 *
+	 * Example:
+	 *
+	 * {{{
```
Convert to javadoc

dawidwys commented on flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/api/java/BatchTableEnvironment.java:
```diff
@@ -0,0 +1,289 @@
+	 * @param name The name under which the {@link DataSet} is registered in the catalog.
+	 * @param dataSet The {@link DataSet} to register.
+	 * @param fields The field names of the registered table.
+	 * @tparam T The type of the {@link DataSet} to register.
```
Convert to `@param <T>`.

Could you please check other javadocs ?

hequn8128 replied:
I will grep and replace them once for all. Sorry for this.

dawidwys commented on flink-table/flink-table-api-java-bridge/src/main/java/org/apache/flink/table/api/java/StreamTableEnvironment.java:
```diff
@@ -0,0 +1,394 @@
+
+	/**
+	 * Converts the given {@link Table} into a {@link DataStream} of add and retract messages.
+	 * The message will be encoded as [[JTuple2]]. The first field is a {@link Boolean} flag,
```
`[[JTuple2]]` to `{@link JTuple2}`

dawidwys commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/tableImpl.scala:
```diff
@@ -393,7 +395,7 @@ class TableImpl(
   override def window(overWindows: OverWindow*): OverWindowedTable = {
 
-    if (tableEnv.isInstanceOf[BatchTableEnvironment]) {
+    if (tableEnv.isInstanceOf[BatchTableEnvImpl]) {
```
What about the java version?

I think it applies to all places we check if it is Batch or Stream.

hequn8128 replied:
What do you mean? The `BatchTableEnvImpl` here is `api.BatchTableEnvImpl`. Both `scala.BatchTableEnvImpl` and `java.BatchTableEnvImpl` extend it.

dawidwys replied:
Sorry, you are right, ignore my comment.

dawidwys commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/descriptors/RegistrableDescriptor.scala:
```diff
@@ -1,49 +0,0 @@
-/**
-  * A trait for descriptors that allow to register table source and/or sinks.
-  */
-trait RegistrableDescriptor extends Descriptor {
```
Why couldn't we move those classes (also e.g. TableFactoryUtil) in the previous commit?

hequn8128 replied:
I will move this change into the right commit.

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/Table.java:
```diff
@@ -89,6 +89,11 @@
+	/**
+	 * Returns the {@link TableEnvironment} of this table.
+	 */
+	TableEnvironment getTableEnvironment();
```
Let's not add this method. My opinion is we should decouple `Table` from `TableEnvironment`. The only obstacle so far is only the `writeToSink` method.

hequn8128 replied:
The conversions in `TableConversions.scala` need TableEnvironment.  I will put `TableConversions` in planner module for now and remove getTableEnvironment() in Table.

hequn8128 commented:
@dawidwys Hi, thank you very much for the review. I have addressed all your comments, except for the one about changing the interface to TableSource/Sink for TableFactories. I also agree with you and it would be nice if we can remove the reflections. However, the main concern is the compatible problem. The interface is widely used by users. Maybe we can have some opinions from @twalthr . What do you think?

Thank you again for the nice review.

Best, Hequn

sunjincheng121 commented:
Thanks for the update @hequn8128!

The changes look good to me! +1 to merged.

Will be merged if there no other comments from @twalthr @dawidwys 

@flinkbot approve all

Best,
Jincheng

dawidwys commented:
Hi @sunjincheng121 ,
I will merge this once I finish the review, are you fine with this?

dawidwys commented:
Merged 32199ed650954d9daa459ad10a1061294431a63d  in ce09ebb815a8c33f9c716b5c01919e8c9041c6a5

dawidwys commented:
Merged e0decce7aa3abba8d697cd1557692e8cd25eb3ec in b6c7ba3aa90669b3dc6aa91e16c7e257f5819ef8

sunjincheng121 commented:
Thank you @dawidwys, sounds good to me, just go ahead!

hequn8128 commented:
@dawidwys Thank you very much for the review and merge. Feel free to let me know if any other problems I need to address. 

Best, Hequn

twalthr commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/Rowtime.java:
```diff
@@ -0,0 +1,140 @@
+
+	private final DescriptorProperties internalProperties = new DescriptorProperties(true);
+
+	public static final String ROWTIME = "rowtime";
```
These static fields should rather go into a `RowtimeValidator` class as they mess up the API class otherwise.

hequn8128 replied:
Make sense. As we can not port `RowtimeValidator` into the common module now, how about extract these properties into a new property class? This not only makes API more clear but also makes these variables more well arranged. What do you think?

twalthr replied:
I'm also fine with keeping the fields there until we ported `RowtimeValidator` but with a TODO comment to move them later.

hequn8128 replied:
Ok, sounds good. I will add a TODO comment. Thank you.

twalthr commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/factories/TableFactoryService.java:
```diff
@@ -0,0 +1,373 @@
+/**
+ * Unified class to search for a {@link TableFactory} of provided type and properties.
+ */
+public class TableFactoryService {
```
This class and factory exceptions should be moved into `table-common` as connectors access it for discovering a format. See the Kafka connector as an example.

hequn8128 replied:
Ok, I will try to move these classes into the table-common module.

twalthr commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/factories/StreamTableSinkFactory.scala:
```diff
@@ -20,15 +20,15 @@ package org.apache.flink.table.factories
   * @tparam T type of records that the factory consumes
   */
-trait StreamTableSinkFactory[T] extends TableFactory {
+trait StreamTableSinkFactory[T] extends TableSinkFactory[T] {
```
Can we also convert this interface to Java now and move them to its final destination?

hequn8128 replied:
I'm not sure. If we put this class into the java-bridge, the scala-bridge needs to depend on java-bridge, because java users also need this class.
Do you think it's right to make scala-bridge depends on java-bridge?

twalthr replied:
We had this discussions in the past when creating FLIP-32. I cite the important parts:
"Currently, we cannot add interfaces for connectors into the common module as classes such as `StreamTableSource` or `BatchTableSource` require a dependency to `DataStream` or `DataSet` API classes. This might change in the future once we reworked our TableSource and TableSink interfaces. For now, extension points for connectors are located in `flink-table-api-*-bridge` and tightly integrate with the target API."

`StreamTableSource` and sink use the Java version of the DataStream API, thus, those two interfaces belong into a Java bridge. If users want to implement a Java source and use the Scala Table API, they can simply add both modules. This is not a problem because classes in both bridging modules are prefixed with `api.java` or `api.scala` or do not have class name clashes.

In the near future we will deprecate those interfaces and have just one unified interface in `common` that does not depend in DataStream API anymore but returns a `SourceFunction`.

wuchong replied:
The `SourceFunction` you mentioned is the new source interface of "FLIP-27: Refactor Source Interface" ? Because the current `SourceFunction` is still in `flink-streaming-java` module.

hequn8128 replied:
@twalthr Thanks for the detailed explanation. I find that scala users need to depend on java-bridge any way either we put this class in planner or in java-bridge. I will port it into the java-bridge. Thank you.

twalthr replied:
@wuchong yes I'm talking about FLIP-27. The new unified `SourceFunction` should be in `flink-core` close to `MapFunction` etc.

twalthr commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/api/StreamTableEnvImpl.scala:
```diff
@@ -159,48 +159,13 @@ abstract class StreamTableEnvironment(
     new StreamTableDescriptor(this, connectorDescriptor)
   }
 
   /**
```
Remove all comments that where defined in the interface. We don't need to maintain consistency in the future.

twalthr commented on flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/api/stream/sql/AggregateTest.scala:
```diff
@@ -64,7 +64,7 @@ class AggregateTest extends TableTestBase {
     streamUtil.addFunction("udag", new MyAgg)
     val aggFunctionDefinition = streamUtil
-      .tableEnv
+      .tableEnv.asInstanceOf[StreamTableEnvImpl]
```
Update the util instead of adding dozens of casts.

twalthr commented on flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/nodes/dataset/DataSetValues.scala:
```diff
@@ -67,7 +67,7 @@ class DataSetValues(
 
   override def translateToPlan(
-      tableEnv: BatchTableEnvironment,
+      tableEnv: BatchTableEnvImpl,
```
Do we still need to pass the implementation through the entire planner? Could you check what is really required in the planner? Maybe we can introduce some `PlannerContext` class instead?

hequn8128 replied:
Totally agree with you. Most of the places only use `tableEnv.getConfig`. However, `tableEnv.execEnv` is used in `DataSetValues`, `DataStreamValues` and `StreamTableSourceScan`.  Also `tableEnv.getRelBuilder` is used in `StreamTableSourceScan`. Maybe we can simplify this a bit or just simply put them into the `PlannerContext`. 

Btw, how about fixing this in another PR. We may need some more discussions. And let this PR focus on the tableEnv interface. What do you think?

twalthr replied:
Totally agree. I thought we are only using `getConfig()`. We can postpone any bigger change here. This code will be dropped anyway soon.

dawidwys commented:
Apart from @twalthr it also does not compile right now. Could you check that also @hequn8128, thank you in advance.

hequn8128 commented:
@dawidwys Sure, sorry about the conflicts. I will try to address the comments from @twalthr  and solve the conflicts by rebasing to the master. Thank you for the notice.

Best, Hequn

hequn8128 commented:
@twalthr @dawidwys Hi, thanks for the review and suggestions. I have updated the PR and rebased to the master. As the first two commits have been merged, the new commits don't include them. I also add a fix commit for the compile error problem. 

BTW, @twalthr I have addressed all your comments except for the one about porting StreamTableSinkFactory into java-bridge. I'm not sure if it is right to let scala-bridge depend on java-bridge. What do you think of it?

Best, Hequn

dawidwys commented:
Hi @hequn8128,
I think the fix(0b33b9472fa48376b70a716e39d35d6e609d1709) is unnecessary, I forgot to commit a change initially, but merged a hotfix right away(62c2a7bda33e17d39d400f0ee40db7d78c6f8484).

Also there are still compile errors e.g. in `org.apache.flink.streaming.connectors.elasticsearch.ElasticsearchUpsertTableSinkFactoryBase`. Could you have a look at it?

hequn8128 commented:
@dawidwys Hi, I have updated the PR. I removed the fix commit which is marked unnecessary by you. I also added a new commit to port table source&sink classes into api-java-bridge module. 
Would be great if you can take another look. Thank you very much. 

Best, Hequn
